controllers/alert_controller.go

```diff
@@ -99,95 +100,80 @@ func (r *AlertReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts Aler
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=alerts/status,verbs=get;update;patch
 
 func (r *AlertReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
-	start := time.Now()
+	reconcileStart := time.Now()
 	log := ctrl.LoggerFrom(ctx)
 
-	alert := &v1beta1.Alert{}
-	if err := r.Get(ctx, req.NamespacedName, alert); err != nil {
+	obj := &apiv1.Alert{}
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// record suspension metrics
-	r.RecordSuspend(ctx, alert, alert.Spec.Suspend)
-
-	if alert.Spec.Suspend {
-		log.Info("Reconciliation is suspended for this object")
-		return ctrl.Result{}, nil
-	}
-
-	patchHelper, err := patch.NewHelper(alert, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	// Initialize the runtime patcher with the current version of the object.
+	patcher := patch.NewSerialPatcher(obj, r.Client)
 
 	defer func() {
-		patchOpts := []patch.Option{
-			patch.WithOwnedConditions{
-				Conditions: []string{
-					meta.ReadyCondition,
-					meta.ReconcilingCondition,
-					meta.StalledCondition,
-				},
-			},
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
+
+		// Emit warning event if the reconciliation failed.
+		if retErr != nil {
+			r.Event(obj, corev1.EventTypeWarning, meta.FailedReason, retErr.Error())
```

Comment:
Since NC is simple, we don't need a helper for logging and eventing yet, but it'd be good to keep them together so that we can easily replace them in the future when we do get a helper for doing both.

Reply:
Oh, nevermind. Just recalled that logging is not needed here because the runtime will log it.
So, the error log in `validate()` can be removed.

Reply:
Okay, nevermind again, just realized the meaning of the comment
> 		// log not found errors since they get filtered out
The error gets filtered in the previous level.